File: Data Structures & Algorithms/house-robber-ii/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    
    robHouse(nums, houseNumber, moneyMap, start) {
        if (houseNumber == nums.length - 1) {
            if(start == 0 && nums.length > 1) {
                return 0
            }
            return nums[houseNumber]
        }
        if (houseNumber >= nums.length) {
            return 0
        }
        if (moneyMap.has(`${houseNumber}|${start}`)) {
            return moneyMap.get(`${houseNumber}|${start}`)
        }
        var result = 0
        const thisHouse = nums[houseNumber] + this.robHouse(nums, houseNumber + 2, moneyMap, start)
        const nextHouse = houseNumber == 0 ? this.robHouse(nums, houseNumber + 1, moneyMap, start + 1) : this.robHouse(nums, houseNumber + 1, moneyMap, start)
        if(thisHouse > nextHouse) {
            result = thisHouse
        } else {
            result = nextHouse
        }
            
        moneyMap.set(`${houseNumber}|${start}`, result)
        return result
    }

    rob(nums) {
        const maxMoney = new Map();
        return this.robHouse(nums, 0, maxMoney, 0)
    }
}

File: Data Structures & Algorithms/jump-game/submission-1.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        return canJumpIndex(0,nums);
    }
    bool canJumpIndex(int index, vector<int>& nums) {
        int length = nums.size();
        if (index == length-1) {
            return true;
        }
        if(index >= length) {
            return false;
        }
        for(int i = nums[index];i>0;i--) {
            if(canJumpIndex(index+i, nums)) {
                return true;
            }
        }
        return false;
    }
};
